tls/s2n_async_pkey.c

```diff
@@ -0,0 +1,346 @@
+
+    int rsa_failed = !!s2n_pkey_decrypt(pkey, encrypted, init_decrypted);
+    GUARD(on_complete(conn, rsa_failed, init_decrypted));
+
+    return 0;
+}
+
+int s2n_async_pkey_sign(struct s2n_connection *conn, s2n_signature_algorithm sig_alg, struct s2n_hash_state *digest,
+                   s2n_async_pkey_sign_complete on_complete)
+{
+    notnull_check(conn);
+    notnull_check(digest);
+    notnull_check(on_complete);
+
+    if (conn->config->async_pkey_cb) {
+        GUARD(s2n_async_pkey_sign_async(conn, sig_alg, digest, on_complete));
+    } else {
+        GUARD(s2n_async_pkey_sign_sync(conn, sig_alg, digest, on_complete));
+    }
+
+    return 0;
+}
+
+int s2n_async_pkey_sign_async(struct s2n_connection *conn, s2n_signature_algorithm sig_alg,
+                              struct s2n_hash_state *digest, s2n_async_pkey_sign_complete on_complete)
+{
+    S2N_ERROR_IF(conn->handshake.async_state != S2N_ASYNC_NOT_INVOKED, S2N_ERR_ASYNC_MORE_THAN_ONE);
+
+    DEFER_CLEANUP(struct s2n_blob mem = { 0 }, s2n_free);
+    GUARD(s2n_alloc(&mem, sizeof(struct s2n_async_pkey_op)));
+    GUARD(s2n_blob_zero(&mem));
+    struct s2n_async_pkey_op *op = ( void * )mem.data;
+
+    op->type = S2N_ASYNC_SIGN;
+    op->conn = conn;
+
+    struct s2n_async_pkey_sign_data *sign = &op->op.sign;
+    sign->on_complete                     = on_complete;
+    sign->sig_alg                         = sig_alg;
+
+    GUARD(s2n_hash_new(&sign->digest));
+    GUARD(s2n_hash_copy(&sign->digest, digest));
+
+    /* Block the handshake and set async state to invoking to block async states */
+    GUARD(s2n_conn_set_handshake_read_block(conn));
+    conn->handshake.async_state = S2N_ASYNC_INVOKING;
+
+    /* async_pkey will own the op, clean the mem blob to avoid freeing op */
+    GUARD(s2n_blob_init(&mem, NULL, 0));
+
+    S2N_ERROR_IF(conn->config->async_pkey_cb(conn, op) < 0, S2N_ERR_ASYNC_CALLBACK_FAILED);
+
+    /* Set state to waiting to allow op to be consumed by connection */
+    conn->handshake.async_state = S2N_ASYNC_INVOKED_WAITING;
+
+    /* Return an async blocked error to drop out of s2n_negotiate loop */
+    S2N_ERROR(S2N_ERR_ASYNC_BLOCKED);
+}
+
+int s2n_async_pkey_sign_sync(struct s2n_connection *conn, s2n_signature_algorithm sig_alg,
+                             struct s2n_hash_state *digest, s2n_async_pkey_sign_complete on_complete)
+{
+    const struct s2n_pkey *pkey = conn->handshake_params.our_chain_and_key->private_key;
+    DEFER_CLEANUP(struct s2n_blob signed_content = { 0 }, s2n_free);
+
+    uint32_t maximum_signature_length = s2n_pkey_size(pkey);
+    GUARD(s2n_alloc(&signed_content, maximum_signature_length));
+
+    GUARD(s2n_pkey_sign(pkey, sig_alg, digest, &signed_content));
+
+    GUARD(on_complete(conn, &signed_content));
+
+    return 0;
+}
+
+int s2n_async_pkey_op_perform(struct s2n_async_pkey_op *op, s2n_cert_private_key *key)
+{
+    notnull_check(op);
+    S2N_ERROR_IF(op->complete, S2N_ERR_ASYNC_ALREADY_PERFORMED);
```

Comment:
`s2n_async_pkey_op_perform` is a part of public s2n.h api and `s2n_result` doesn't seem to be exposed there. Are we planning on exposing it? I can also create a wrapper for each function, where internal ones will use `S2N_RESULT`, and external one will convert it to posix, but that seems to be a bit redundant.

Reply:
`ENSURE_POSIX` is also available if you're not wanting to use `S2N_RESULT`

Reply:
I see, will update the code to `ENSURE`. For my understanding:
1. Is `*_POSIX` supposed to be used by functions in external API or is it ok to use it internally too?
2. Right now there is no difference in the end result of using `BAIL_POSIX` and `S2N_ERROR`, right? Is there going to be in a future? Are we deprecating the later?

Reply:
1. It can be used anywhere that is interacting with a function that's returning an `int` where `0` on success and `-1`  on failure. It's totally fine to use it for both internal and external functions that fit that description. The public s2n API is, unfortunately, not going to return s2n_result so it's inevitable that we still have _POSIX macros for the foreseeable future. But internally, the intent is to move as much as possible to use s2n_result, as it's a huge win for correctness (see #1782 and #1783).
2. Correct. The S2N_ERROR macro will be deprecated as well. I think there are a couple of issues with the name: it's a noun which is a bit confusing. I initially thought it was creating an error and the function was in charge of control flow. BAIL, on the other hand, is a verb and indicates that the function will be returning early. Also the S2N_ prefix isn't really needed and is inconsistent with GUARD and other safety macros.